src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.Unix.cs

```diff
@@ -0,0 +1,186 @@
+
+            nuint adjustedAlignment = Math.Max(alignment, (uint)sizeof(void*));
+            void* result = Interop.Sys.AlignedAlloc(adjustedAlignment, (byteCount != 0) ? (byteCount + (adjustedAlignment - 1)) & ~(adjustedAlignment - 1) : adjustedAlignment);
+
+            if (result == null)
+            {
+                ThrowHelper.ThrowOutOfMemoryException();
+            }
+
+            return result;
+        }
+
+        /// <summary>Frees an aligned block of memory.</summary>
+        /// <param name="ptr">A pointer to the aligned block of memory that should be freed.</param>
+        /// <remarks>
+        ///    <para>This method does nothing if <paramref name="ptr" /> is <c>null</c>.</para>
+        ///    <para>This method is a thin wrapper over the C <c>free</c> API or a platform dependent aligned free API such as <c>_aligned_free</c> on Win32.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void AlignedFree(void* ptr)
+        {
+            if (ptr != null)
+            {
+                Interop.Sys.AlignedFree(ptr);
+            }
+        }
+
+        /// <summary>Reallocates an aligned block of memory of the specified size and alignment, in bytes.</summary>
+        /// <param name="ptr">The previously allocated block of memory.</param>
+        /// <param name="byteCount">The size, in bytes, of the block to allocate.</param>
+        /// <param name="alignment">The alignment, in bytes, of the block to allocate. This must be a power of <c>2</c>.</param>
+        /// <returns>A pointer to the reallocated aligned block of memory.</returns>
+        /// <exception cref="ArgumentException"><paramref name="alignment" /> is not a power of two.</exception>
+        /// <exception cref="OutOfMemoryException">Reallocating <paramref name="byteCount" /> of memory with <paramref name="alignment" /> failed.</exception>
+        /// <remarks>
+        ///     <para>This method acts as <see cref="AlignedAlloc" /> if <paramref name="ptr" /> is <c>null</c>.</para>
+        ///     <para>This method allows <paramref name="byteCount" /> to be <c>0</c> and will return a valid pointer that should not be dereferenced and that should be passed to free to avoid memory leaks.</para>
+        ///     <para>This method is a platform dependent aligned reallocation API such as <c>_aligned_realloc</c> on Win32.</para>
+        ///     <para>This method is not compatible with <see cref="Free" /> or <see cref="Realloc" />, instead <see cref="AlignedFree" /> or <see cref="AlignedRealloc" /> should be called.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* AlignedRealloc(void* ptr, nuint byteCount, nuint alignment)
+        {
+            void* newPtr = AlignedAlloc(byteCount, alignment);
+
+            if (ptr != null)
+            {
+                Buffer.Memmove(ref *(byte*)newPtr, ref *(byte*)ptr, byteCount);
```

Comment:
I do not think this works. `byteCount` is the new size, but we can only copy old size (the rest should be left uninitialized).

Reply:
Ah right. I'll need to either expose `malloc_usable_size` or make this a P/Invoke to a wrapper `SystemNative.AlignedRealloc`

Reply:
Updated to be a P/Invoke to a `SystemNative_AlignedRealloc` to avoid needing 3-4 P/Invoke transitions.

It now correctly does:
* Allocate new space
* If allocation was successful:
  * Get usable size of previous allocation
  * Copy min of old/new size from old to new
  * Free old allocation
* Return result

I've also added a test that explicitly tests going from a small to large allocation and that the bytes are as expected.